dbuf/dbuf/client.py

```diff
@@ -0,0 +1,171 @@
+        if rng is None:
+            rng = random.Random()
+        self.rng = rng
+        self.deploy_config = deploy_config
+        self.root_url = deploy_config.base_url(leader_name)
+        self.session_url = None if id is None else f'{self.root_url}/s/{id}'
+        self.aiosession = None
+        self.id = id
+        self.offs = []
+        self.sizes = []
+        self.cursor = 0
+        self.max_bufsize = max_bufsize
+
+    async def create(self):
+        resp = await utils.request_retry_transient_errors(
+            self.aiosession,
+            'POST',
+            f'{self.root_url}/s')
+        assert resp.status == 200
+        self.id = int(await resp.text())
+        self.session_url = f'{self.root_url}/s/{self.id}'
+        return self.id
+
+    async def start_write(self):
+        workers = await self.get_workers()
+        worker = workers[self.rng.randrange(len(workers))]
+        return DBufAppender(worker, self.deploy_config, self.id, self.max_bufsize, self.aiosession)
+
+    async def get(self, key):
+        server = key[0]
+        server_url = self.deploy_config.base_url(server)
+
+        resp = await utils.request_retry_transient_errors(
+            self.aiosession,
+            'POST',
+            f'{server_url}/s/{self.id}/get',
+            json=key)
+        assert resp.status == 200
+        return await resp.read()
+
+    def _decode(self, byte_array):
+        off = 0
+        result = []
+        while off < len(byte_array):
+            n2 = struct.unpack_from("i", byte_array, off)[0]
+            off += 4
+            result.append(byte_array[off:(off+n2)])
+            off += n2
+        return result
+
+    async def getmany(self, keys, retry_delay=1):
+        servers = collections.defaultdict(list)
+        results = [None for _ in keys]
+        for i, key in enumerate(keys):
+            servers[key[0]].append((key, i))
+
+        async def get_from_server(server, keys):
+            server_url = self.deploy_config.base_url(server)
+            i = 0
+            while i < len(keys):
+                batch = []
+                size = 0
+                while i < len(keys):
+                    assert keys[i][0][3] < self.max_bufsize
+                    if size + keys[i][0][3] < self.max_bufsize:
+                        batch.append(keys[i])
+                        size += keys[i][0][3]
+                        i += 1
+                    else:
+                        break
+
+                resp = await utils.request_retry_transient_errors(
+                    self.aiosession,
+                    'POST',
+                    f'{server_url}/s/{self.id}/getmany',
+                    json=[x[0] for x in batch])
+                assert resp.status == 200
+                data = await resp.read()
+                for v, j in zip(self._decode(data), (x[1] for x in batch)):
+                    results[j] = v
+        await asyncio.gather(*[get_from_server(server, keys)
```

Comment:
Unbounded gathers are almost certainly not what you want.  hailtop.utils has a bounded_gather.  We should always be using that unless we know gather is suitably bounded.

Reply:
How do you define "suitable bounded"?